src/cryptography/hazmat/primitives/kdf/kbkdf.py

```diff
@@ -0,0 +1,160 @@
+
+        if not self._valid_byte_length(llen):
+            raise ValueError('llen must be between 1 and 4')
+
+        self._algorithm = algorithm
+        self._mode = mode
+        self._length = length
+        self._rlen = rlen
+        self._llen = llen
+        self._location = location
+        self._label = label
+        self._context = context
+        self._backend = backend
+        self._used = False
+        self._fixed_data = fixed
+
+    def _valid_byte_length(self, value):
+        if not isinstance(value, int):
+            raise TypeError('value must be of type int')
+
+        if not 1 <= value <= 4:
+            return False
+        return True
+
+    def _int_as_binary_representation(self, length, value):
+        if not self._valid_byte_length(length):
+            raise ValueError('binary length must be between 1 and 4')
+
+        if length == 1:
+            return struct.pack(">B", value)
+        elif length == 2:
+            return struct.pack(">H", value)
+        elif length == 3:
+            return struct.pack(">L", value)[1:]
+        else:
+            return struct.pack(">L", value)
+
+    def derive(self, key_material):
+        if self._used:
+            raise AlreadyFinalized
+
+        if not isinstance(key_material, bytes):
+            raise TypeError('key_material must be bytes')
+        self._used = True
+
+        # inverse floor division (equivalent to ceiling)
+        rounds = -(-self._length // self._algorithm.digest_size)
+
+        output = [b'']
+        r_bin_len = self._int_as_binary_representation(self._rlen, 1)
+        r_bin_rep = binascii.hexlify(r_bin_len)
+        if rounds > pow(2, min(len(r_bin_rep), 32)) - 1:
+            raise ValueError('There are too many iterations.')
+
+        for i in range(1, rounds + 1):
+            h = hmac.HMAC(key_material, self._algorithm, backend=self._backend)
+            counter = self._int_as_binary_representation(self._rlen, i)
+            if self._location == KBKDFHMAC.LOCATION_BEFORE_FIXED:
+                h.update(counter)
+
+            h.update(self._generate_fixed_input())
+
+            if self._location == KBKDFHMAC.LOCATION_AFTER_FIXED:
+                h.update(counter)
+
+            output.append(h.finalize())
+
+        return b''.join(output)[:self._length]
+
+    def _generate_fixed_input(self):
+        """
+        Combine the fixed data (label and context) to a binary string
+
+        :return: binary string
+        """
+        if self._fixed_data and isinstance(self._fixed_data, bytes):
+            # NIST's test vectors are only supply Fixed input data
+            return self._fixed_data
+
+        l = self._int_as_binary_representation(self._llen, self._length * 8)
+
+        fixed_input = list()
+        fixed_input.append(self._label)
+        fixed_input.append(six.int2byte(0))
+        fixed_input.append(self._context)
+        fixed_input.append(l)
+        return b''.join(fixed_input)
```

Comment:
I don't believe you can do string formatting on a byte object. `%` formatting to byte objects were added in [version 3.5](https://docs.python.org/3/whatsnew/3.5.html#pep-461-formatting-support-for-bytes-and-bytearray), but it doesn't mention anything about `format()`. 

If you prefer a more elegant one liner, I can change `return b''.join(fixed_input)` to `return b''.join([self._label, b"x\00", self._context, l])`